### ISSUE TITLE
How to run your code?

### ISSUE THREAD

**Issue by iganggang** [RogerNi/MoLE#4]
Thank u very much for your interesting work.

1）i put run_all_3_seeds.sh in the same directory as run_longExp.py, and right click run run_all_3_seeds.sh. but it shows 
but it shows many problem.
such as 

![Image](https://github.com/user-attachments/assets/d83e6e7e-a2c9-444a-b389-7285b0e60495)

![Image](https://github.com/user-attachments/assets/6df8cc42-b10c-4c1e-a4ea-24afb07c1519)

2）how to run run_longExp.py?

![Image](https://github.com/user-attachments/assets/ef1e7bb8-a5c4-46b8-933d-05f1722f6b21)

could u help me to solve it?

**Comment by RogerNi** [comment]
If you are using `run_all_3_seeds.sh`, you should be in the same directory as that file, which is `MoLE/scripts`. Simply copying everything to the root directory of the repository (`MoLE`) may not work, as some paths are relative to the original script locations.

If you prefer to use `run_longExp.py` directly, you need to provide the appropriate parameters. Please refer to the `MoLE/scripts/MoE_*.sh` scripts to see which parameters are required.

**Comment by iganggang** [comment]
Thank you for your reply.
1) i run the run_longExp.py. it has three resutls.

![Image](https://github.com/user-attachments/assets/c62295af-a842-4739-ae8b-4c2bc5864168)
i use the third results. But i find the mse=0.412394, is smaller than 0.453 in the paper. such as the figure shows, 

![Image](https://github.com/user-attachments/assets/1d31c292-e57b-4702-9eb1-aafcfd45260d)
i do not revise any hyperparameters，just in the exp_main file, revise drop_last=True -> False, 
`        # preds = np.array(preds)
        # trues = np.array(trues)
        # inputx = np.array(inputx)

        preds = np.concatenate(preds, axis=0)
        trues = np.concatenate(trues, axis=0)
        inputx = np.concatenate(inputx, axis=0)
`
why the results are better？
Could you help me to see where is the error happens?

**Comment by RogerNi** [comment]
I don’t think there’s any issue with the way you’re executing the code. However, the discrepancy in results could be due to several reasons:
	1.	The results shown in the table are not the best test metrics achieved overall, but rather the test metrics corresponding to the lowest validation loss. That is, the validation set was used to select the best hyperparameter configuration, and the corresponding test result under that configuration is what’s reported.
	2.	The software environment may differ from the one I used, including library versions.
	3.	PyTorch does not guarantee reproducibility across devices. Different hardware (e.g., GPU accelerators) can lead to slight variations in results.

That said, if you’re conducting a full hyperparameter grid search, the variability introduced by reasons 2 and 3 should average out, and you should still observe similar overall performance.

**Comment by iganggang** [comment]
Thank you for your rapid reply.
if I want to reproduce the results in your paper, how do I do?

**Comment by RogerNi** [comment]
You can use `scripts/run_all_3_seeds.sh` to rerun all experiments from the paper. For the main results, we use only the random seed 2021, following prior work. You may modify the script accordingly to use only this seed. While the exact numbers may vary, the overall performance should be comparable.

**Comment by RogerNi** [comment]
Also, I’m not quite sure I understand your concern:
>your results are so good in the EXP that I can not put them on paper.

Could you explain more?

**Comment by iganggang** [comment]
And, i try to run the `run_all_3_seeds.sh`, the result.txt is too long to read. and after one day, the code is still running at predict_length=96. so i want to try run by dataset. then i run` MoE_ETTs.sh`, but the file directly make me confused. 
`head_dropout=$5
ett=$4
(
pred_len=$1
lr=$2
t_dim=$3
aug_rate=0
aug_method=f_mask
`
Could u write some simple code to run? like pred_len=96?

**Comment by iganggang** [comment]
> Also, I’m not quite sure I understand your concern:
> 
> > your results are so good in the EXP that I can not put them on paper.
> 
> Could you explain more?

i run etth1 336 predict 96 by using MoLE_DLinear, but my results is 0.412 is better than 0.453 in your paper. 

![Image](https://github.com/user-attachments/assets/057939f9-e350-40e8-8669-cd6da560856e)
DLinear 0.412, MoLE_DLinear 0.453, i run MoLE_DLinear is 0.412
Your results are even better in the paper when I reproduce your experiments.

**Comment by RogerNi** [comment]
I understand you may not be familiar with the scripts, or even how to run them, but please note that this project is no longer actively maintained—only bugs will be fixed moving forward. The codebase has already been simplified, and its structure, complexity, and interpretability are on par with related work.

Regarding your specific question:
```
head_dropout=$5
ett=$4
(
pred_len=$1
lr=$2
t_dim=$3
```
These lines simply assign the script’s input arguments to variables used as hyperparameters for the experiments. I’m not sure which part is causing confusion—if you can be more specific, I’d be happy to clarify.

As for the discrepancy in results: if you are running the experiments with only a single set of hyperparameters, then you’re not replicating the main results shown in the paper. As I mentioned earlier:
> The results shown in the table are not the best test metrics achieved overall, but rather the test metrics corresponding to the lowest validation loss. That is, the validation set was used to select the best hyperparameter configuration, and the corresponding test result under that configuration is what’s reported.

To reproduce the reported results, you may want to perform a hyperparameter grid search (like what `run_all_3_seeds.sh` does).

**Comment by iganggang** [comment]
Sorry for later raply. I am stuff.
1) I feel confused, when run `run_longExp.py`, during the training, according to the valid loss to confirm the next epoch training. why in hyperparameter searching, still use valid loss, not test loss as  metrics, is it common phenomenon everyone do that? i think hyperparameter searching to get min test loss.
2) the `run_all_3_seeds.sh` is still runing after 3 days, and the `resutls.txt ` are huge. i am anxiety. So, i want to do run by dataset and dataset. so i try to run `MoE_ETTs.sh`, then there are some parameters `head_dropout=$5
ett=$4
(
pred_len=$1
lr=$2
t_dim=$3`
why i change pred_len=96, it shows errors. `run_longExp.py: error: argument --pred_len: expected one argument`.
i recently see `ETT.sh` like this.

```bash
#!/bin/bash
if [ ! -d "./logs" ]; then
    mkdir ./logs
fi

if [ ! -d "./logs/LongForecasting" ]; then
    mkdir ./logs/LongForecasting
fi
seq_len=336
model_name=DLinear

python -u run_longExp.py \
  --is_training 1 \
  --root_path ./data/ETT/ \
  --data_path ETTh1.csv \
  --model_id ETTh1_$seq_len'_'96 \
  --model $model_name \
  --data ETTh1 \
  --features M \
  --seq_len $seq_len \
  --pred_len 96 \
  --enc_in 7 \
  --des 'Exp' \
  --itr 1 --batch_size 32 --learning_rate 0.005>logs/LongForecasting/$model_name'_'Etth1_$seq_len'_'96.log

python -u run_longExp.py \
  --is_training 1 \
  --root_path ./data/ETT/ \
  --data_path ETTh1.csv \
  --model_id ETTh1_$seq_len'_'192 \
  --model $model_name \
  --data ETTh1 \
  --features M \
  --seq_len $seq_len \
  --pred_len 192 \
  --enc_in 7 \
  --des 'Exp' \
  --itr 1 --batch_size 32 --learning_rate 0.005 >logs/LongForecasting/$model_name'_'Etth1_$seq_len'_'192.log

python -u run_longExp.py \
  --is_training 1 \
  --root_path ./data/ETT/ \
  --data_path ETTh1.csv \
  --model_id ETTh1_$seq_len'_'336 \
  --model $model_name \
  --data ETTh1 \
  --features M \
  --seq_len $seq_len \
  --pred_len 336 \
  --enc_in 7 \
  --des 'Exp' \
  --itr 1 --batch_size 32 --learning_rate 0.005 >logs/LongForecasting/$model_name'_'Etth1_$seq_len'_'336.log

python -u run_longExp.py \
  --is_training 1 \
  --root_path ./data/ETT/ \
  --data_path ETTh1.csv \
  --model_id ETTh1_$seq_len'_'720 \
  --model $model_name \
  --data ETTh1 \
  --features M \
  --seq_len $seq_len \
  --pred_len 720 \
  --enc_in 7 \
  --des 'Exp' \
  --itr 1 --batch_size 32 --learning_rate 0.005 >logs/LongForecasting/$model_name'_'Etth1_$seq_len'_'720.log

**Comment by RogerNi** [comment]
1. Not every work performs hyperparameter tuning. Some simply stick to a single set of hyperparameters without exploring other combinations. This can lead to the issue of hyperparameters being overfitted to the test split of the dataset, aiming to achieve the best test performance. A few studies address this by running the model under multiple random seeds and evaluating whether the model consistently outperforms baseline models from a statistical perspective. In this work, since many hyperparameters cannot be determined globally, a separate hyperparameter search is conducted for each experiment. The validation set is used to select hyperparameters in order to avoid overfitting to the test split. Reporting the lowest test metric may look better, but it is not a fair way to represent our model's performance in a more general setting.

2. If you would like to use `MoE_ETTs.sh` to run a single experiment with `pred_len=96`, you can use the command  
`SEED=$seed MODEL=$model $script_full_path/MoE_ETTs.sh $pred_len $lr $t_dim $ett $head_dropout`,  
just as shown in `run_all_3_seeds.sh`.  
For example, a complete command would be:  
`SEED=2021 MODEL=DLinear ./MoE_ETTs.sh 96 0.005 1 h1 0`.

**Comment by RogerNi** [comment]
Closing due to inactivity.
